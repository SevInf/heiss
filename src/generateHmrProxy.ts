import { parse } from 'acorn';
import {
    ExportNamedDeclaration,
    FunctionDeclaration,
    VariableDeclaration,
    ClassDeclaration,
    VariableDeclarator
} from 'estree';

class HMRProxyGenerator {
    private source: string;
    private originalUrl: string;

    // map from export name to alias
    private proxiedExports: Map<string, string> = new Map();
    private usedNames: Set<string> = new Set();

    constructor(source: string, originalUrl: string) {
        this.source = source;
        this.originalUrl = originalUrl;
    }

    generate(): string {
        const program = parse(this.source, { sourceType: 'module' });
        for (const statement of program.body) {
            if (statement.type === 'ExportNamedDeclaration') {
                this.proxyNamedDeclaration(statement);
            } else if (statement.type === 'ExportDefaultDeclaration') {
                this.proxyName('default', '_default');
            }
        }

        if (this.proxiedExports.size === 0) {
            return this.source;
        }
        return this.generateProxyModule();
    }

    private proxyNamedDeclaration(exportDeclaration: ExportNamedDeclaration) {
        if (exportDeclaration.declaration) {
            this.proxyDeclaration(exportDeclaration.declaration);
        }
        for (const specifier of exportDeclaration.specifiers) {
            this.proxyName(specifier.exported.name);
        }
    }

    private proxyDeclaration(declaration: VariableDeclaration | FunctionDeclaration | ClassDeclaration) {
        if (declaration.type === 'VariableDeclaration') {
            for (const variableDeclaration of declaration.declarations) {
                this.proxyVariableDeclarator(variableDeclaration);
            }
        } else {
            // class or function
            this.proxyName(declaration.id!.name);
        }
    }

    private proxyVariableDeclarator(variable: VariableDeclarator) {
        switch (variable.id.type) {
            case 'Identifier':
                this.proxyName(variable.id.name);
                break;
            // TODO: destructuring
        }
    }

    private proxyName(name: string, alias: string = name) {
        if (this.usedNames.has(alias)) {
            alias = this.getUniqueName(alias);
        }
        this.usedNames.add(alias);
        this.proxiedExports.set(name, alias);
    }

    private generateProxyModule(): string {
        const exportNames = [];
        const imports = [];
        const assignments = [];
        const reexports = [];
        const reassignments = [];
        let clientVarName = this.getUniqueName('client');

        for (const [proxiedName, importAlias] of this.proxiedExports) {
            exportNames.push(proxiedName);
            if (proxiedName === importAlias) {
                imports.push(`    ${proxiedName}`);
            } else {
                imports.push(`    ${proxiedName} as ${importAlias}`);
            }

            const localVariableName = this.getUniqueName(importAlias);
            assignments.push(`let ${localVariableName} = ${importAlias};`);
            reexports.push(`    ${localVariableName} as ${proxiedName}`);
            reassignments.push(`        ${localVariableName} = updated.${proxiedName};`);
        }

        return [
            'import {',
            imports.join(',\n'),
            `} from '${this.originalUrl}?mtime=0';`,
            `import { ${clientVarName} } from "/@hmr";`,
            assignments.join('\n'),
            '',
            'export {',
            reexports.join(',\n'),
            '};',
            '',
            `${clientVarName}.registerModule(`,
            `    "${this.originalUrl}",`,
            `    ${JSON.stringify(exportNames)},`,
            `    (updated) => {`,
            reassignments.join('\n'),
            '});'
        ].join('\n');
    }

    private getUniqueName(desiredName: string): string {
        if (!this.usedNames.has(desiredName)) {
            this.usedNames.add(desiredName);
            return desiredName;
        }

        let idx = 0;
        let name;
        do {
            name = `${desiredName}${idx++}`;
        } while (this.usedNames.has(name));

        this.usedNames.add(name);
        return name;
    }
}

export function generateHmrProxy(source: string, originalUrl: string): string {
    const generator = new HMRProxyGenerator(source, originalUrl);
    return generator.generate();
}
