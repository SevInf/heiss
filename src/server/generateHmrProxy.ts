import { URL } from 'url';
import { parseModule } from './parseModule';
import {
    ExportNamedDeclaration,
    FunctionDeclaration,
    VariableDeclaration,
    ClassDeclaration,
    VariableDeclarator
} from 'estree';

class HMRProxyGenerator {
    private source: string;
    private originalUrl: URL;

    // map from export name to alias
    private proxiedExports: Map<string, string> = new Map();
    private usedNames: Set<string> = new Set();
    private imports: Set<string> = new Set();

    constructor(source: string, originalUrl: URL) {
        this.source = source;
        this.originalUrl = originalUrl;
    }

    generate(): string {
        const program = parseModule(this.source);
        for (const statement of program.body) {
            switch (statement.type) {
                case 'ExportNamedDeclaration':
                    this.proxyNamedDeclaration(statement);
                    break;
                case 'ExportDefaultDeclaration':
                    this.proxyName('default', '_default');
                    break;
                case 'ImportDeclaration':
                    this.imports.add(this.resolveImport(statement.source.value as string));
                    break;
            }
        }

        if (this.proxiedExports.size === 0) {
            return this.generateNoExportsProxy();
        }
        return this.generateExportsProxy();
    }

    private resolveImport(relativeImport: string): string {
        return new URL(relativeImport, this.originalUrl).href;
    }

    private proxyNamedDeclaration(exportDeclaration: ExportNamedDeclaration) {
        if (exportDeclaration.declaration) {
            this.proxyDeclaration(exportDeclaration.declaration);
        }
        for (const specifier of exportDeclaration.specifiers) {
            this.proxyName(specifier.exported.name);
        }
    }

    private proxyDeclaration(declaration: VariableDeclaration | FunctionDeclaration | ClassDeclaration) {
        if (declaration.type === 'VariableDeclaration') {
            for (const variableDeclaration of declaration.declarations) {
                this.proxyVariableDeclarator(variableDeclaration);
            }
        } else {
            // class or function
            this.proxyName(declaration.id!.name);
        }
    }

    private proxyVariableDeclarator(variable: VariableDeclarator) {
        switch (variable.id.type) {
            case 'Identifier':
                this.proxyName(variable.id.name);
                break;
            // TODO: destructuring
        }
    }

    private proxyName(name: string, alias: string = name) {
        if (this.usedNames.has(alias)) {
            alias = this.getUniqueName(alias);
        }
        this.usedNames.add(alias);
        this.proxiedExports.set(name, alias);
    }

    private generateNoExportsProxy(): string {
        return [
            `import "${this.originalUrl}?mtime=0";`,
            `import { client } from "/@hmr";`,
            '',
            'client.registerModule(',
            `    "${this.originalUrl}",`,
            '    [],',
            `    ${JSON.stringify(Array.from(this.imports))}`,
            ');'
        ].join('\n');
    }

    private generateExportsProxy(): string {
        const exportNames = [];
        const imports = [];
        const assignments = [];
        const reexports = [];
        const reassignments = [];
        let clientVarName = this.getUniqueName('client');

        for (const [proxiedName, importAlias] of this.proxiedExports) {
            exportNames.push(proxiedName);
            if (proxiedName === importAlias) {
                imports.push(`    ${proxiedName}`);
            } else {
                imports.push(`    ${proxiedName} as ${importAlias}`);
            }

            const localVariableName = this.getUniqueName(importAlias);
            assignments.push(`let ${localVariableName} = ${importAlias};`);
            reexports.push(`    ${localVariableName} as ${proxiedName}`);
            reassignments.push(`        ${localVariableName} = updated.${proxiedName};`);
        }

        return [
            'import {',
            imports.join(',\n'),
            `} from '${this.originalUrl}?mtime=0';`,
            `import { ${clientVarName} } from "/@hmr";`,
            assignments.join('\n'),
            '',
            'export {',
            reexports.join(',\n'),
            '};',
            '',
            `${clientVarName}.registerModule(`,
            `    ${JSON.stringify(this.originalUrl)},`,
            `    ${JSON.stringify(exportNames)},`,
            `    ${JSON.stringify(Array.from(this.imports))},`,
            `    (updated) => {`,
            reassignments.join('\n'),
            '    }',
            ');'
        ].join('\n');
    }

    private getUniqueName(desiredName: string): string {
        if (!this.usedNames.has(desiredName)) {
            this.usedNames.add(desiredName);
            return desiredName;
        }

        let idx = 0;
        let name;
        do {
            name = `${desiredName}${idx++}`;
        } while (this.usedNames.has(name));

        this.usedNames.add(name);
        return name;
    }
}

export function generateHmrProxy(source: string, originalUrl: URL): string {
    const generator = new HMRProxyGenerator(source, originalUrl);
    return generator.generate();
}
