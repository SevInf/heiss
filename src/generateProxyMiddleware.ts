import fs from 'fs-extra';
import * as path from 'path';
import send from 'koa-send';
import { Middleware } from 'koa';
import { generateHmrProxy } from './generateHmrProxy';

interface Options {
    rootPath: string;
}
export function generateProxyMiddleware(options: Options): Middleware {
    return async (context, next) => {
        const requestPath = context.request.path;
        if (!/\.(js|mjs)$/.test(requestPath)) {
            return next();
        }
        const filePath = path.join(options.rootPath, requestPath);

        if (context.request.query.mtime) {
            await send(context, requestPath, {
                root: options.rootPath
            });
            return;
        }

        let content;
        try {
            content = await fs.readFile(filePath, 'utf8');
        } catch (e) {
            if (e.code === 'ENOENT') {
                return next();
            }
            throw e;
        }

        context.response.set('Content-Type', 'application/javascript');
        context.response.body = generateHmrProxy(content, context.request.url);
    };
}
