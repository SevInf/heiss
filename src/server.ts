'use strict';

import * as fs from 'fs-extra';
import * as path from 'path';
import sane = require('sane');
import Koa from 'koa';
import route from 'koa-route';
import koaStatic from 'koa-static';
import koaWebsocket = require('koa-websocket');
import * as WebSocket from 'ws';
import { generateProxyMiddleware } from './generateProxyMiddleware';

const enum MessageType {
    CHANGE = 'change'
}

interface Message {
    type: MessageType.CHANGE;
    url: string;
    mtime: number;
}

const examplePath = path.resolve(__dirname, '..', 'example');
const clientPath = path.resolve(__dirname, '..', 'client', 'client.mjs');

class Server {
    private sockets: Set<WebSocket>;
    private app: koaWebsocket.App;

    constructor() {
        this.sockets = new Set();
        this.app = koaWebsocket(new Koa());
    }

    async start() {
        this.app.use(
            route.get('/@hmr', context => {
                context.response.type = 'application/javascript';
                context.response.body = fs.createReadStream(clientPath);
            })
        );
        this.app.ws.use(
            route.all('/@hmr/socket', context => {
                const websocket = (context as koaWebsocket.MiddlewareContext).websocket;
                this.sockets.add(websocket);
                websocket.on('close', () => {
                    console.log('socket disconnected');
                    this.sockets.delete(websocket);
                });
            })
        );
        this.app.use(generateProxyMiddleware({ rootPath: examplePath }));
        this.app.use(koaStatic(examplePath));
        await this.app.listen(8080);

        const watcher = sane(examplePath);
        watcher.on('change', (filepath, root, stat) => {
            this.broadcast({
                type: MessageType.CHANGE,
                // TODO: generate from config
                url: `http://localhost:8080/${filepath}`,
                mtime: stat.mtimeMs
            });
        });
    }

    broadcast(message: Message) {
        const stringified = JSON.stringify(message);
        for (const socket of this.sockets) {
            socket.send(stringified);
        }
    }
}

const server = new Server();
server.start();
