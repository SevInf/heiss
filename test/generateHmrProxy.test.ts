import { generateHmrProxy } from '../src/generateHmrProxy';
import { readdirSync, readFileSync } from 'fs';
import * as path from 'path';

describe('generateHmrProxy', () => {
    const fixturesBasePath = path.join(__dirname, 'fixtures', 'generateHmrProxy');
    const fixtures = readdirSync(fixturesBasePath);

    for (const fixture of fixtures) {
        test(path.basename(fixture, '.mjs'), () => {
            const source = readFileSync(path.join(fixturesBasePath, fixture), 'utf8');
            expect(generateHmrProxy(source, `/${fixture}`)).toMatchSnapshot();
        });
    }
})