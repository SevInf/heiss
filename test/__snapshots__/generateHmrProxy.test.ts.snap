// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateHmrProxy client-named-export 1`] = `
"import {
    client
} from 'http://example.com/client-named-export.mjs?mtime=0';
import { client0 } from \\"/@hmr\\";
let client1 = client;

export {
    client1 as client
};

client0.registerModule(
    \\"http://example.com/client-named-export.mjs\\",
    [\\"client\\"],
    [],
    (updated) => {
        client1 = updated.client;
    }
);"
`;

exports[`generateHmrProxy default-export 1`] = `
"import {
    default as _default
} from 'http://example.com/default-export.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let _default0 = _default;

export {
    _default0 as default
};

client.registerModule(
    \\"http://example.com/default-export.mjs\\",
    [\\"default\\"],
    [],
    (updated) => {
        _default0 = updated.default;
    }
);"
`;

exports[`generateHmrProxy default-export-conflict 1`] = `
"import {
    default as _default,
    _default as _default0
} from 'http://example.com/default-export-conflict.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let _default1 = _default;
let _default00 = _default0;

export {
    _default1 as default,
    _default00 as _default
};

client.registerModule(
    \\"http://example.com/default-export-conflict.mjs\\",
    [\\"default\\",\\"_default\\"],
    [],
    (updated) => {
        _default1 = updated.default;
        _default00 = updated._default;
    }
);"
`;

exports[`generateHmrProxy destructuring-array 1`] = `
"import {
    a,
    b
} from 'http://example.com/destructuring-array.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;
let b0 = b;

export {
    a0 as a,
    b0 as b
};

client.registerModule(
    \\"http://example.com/destructuring-array.mjs\\",
    [\\"a\\",\\"b\\"],
    [],
    (updated) => {
        a0 = updated.a;
        b0 = updated.b;
    }
);"
`;

exports[`generateHmrProxy destructuring-array-rest 1`] = `
"import {
    a,
    rest
} from 'http://example.com/destructuring-array-rest.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;
let rest0 = rest;

export {
    a0 as a,
    rest0 as rest
};

client.registerModule(
    \\"http://example.com/destructuring-array-rest.mjs\\",
    [\\"a\\",\\"rest\\"],
    [],
    (updated) => {
        a0 = updated.a;
        rest0 = updated.rest;
    }
);"
`;

exports[`generateHmrProxy destructuring-nested 1`] = `
"import {
    export1,
    export2,
    export3,
    export4
} from 'http://example.com/destructuring-nested.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let export10 = export1;
let export20 = export2;
let export30 = export3;
let export40 = export4;

export {
    export10 as export1,
    export20 as export2,
    export30 as export3,
    export40 as export4
};

client.registerModule(
    \\"http://example.com/destructuring-nested.mjs\\",
    [\\"export1\\",\\"export2\\",\\"export3\\",\\"export4\\"],
    [],
    (updated) => {
        export10 = updated.export1;
        export20 = updated.export2;
        export30 = updated.export3;
        export40 = updated.export4;
    }
);"
`;

exports[`generateHmrProxy destructuring-object 1`] = `
"import {
    a,
    b
} from 'http://example.com/destructuring-object.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;
let b0 = b;

export {
    a0 as a,
    b0 as b
};

client.registerModule(
    \\"http://example.com/destructuring-object.mjs\\",
    [\\"a\\",\\"b\\"],
    [],
    (updated) => {
        a0 = updated.a;
        b0 = updated.b;
    }
);"
`;

exports[`generateHmrProxy destructuring-object-aliases 1`] = `
"import {
    exportName
} from 'http://example.com/destructuring-object-aliases.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let exportName0 = exportName;

export {
    exportName0 as exportName
};

client.registerModule(
    \\"http://example.com/destructuring-object-aliases.mjs\\",
    [\\"exportName\\"],
    [],
    (updated) => {
        exportName0 = updated.exportName;
    }
);"
`;

exports[`generateHmrProxy destructuring-object-default 1`] = `
"import {
    a
} from 'http://example.com/destructuring-object-default.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;

export {
    a0 as a
};

client.registerModule(
    \\"http://example.com/destructuring-object-default.mjs\\",
    [\\"a\\"],
    [],
    (updated) => {
        a0 = updated.a;
    }
);"
`;

exports[`generateHmrProxy dynamic-imports 1`] = `
"import \\"http://example.com/dynamic-imports.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerModule(
    \\"http://example.com/dynamic-imports.mjs\\",
    [],
    []
);"
`;

exports[`generateHmrProxy import-meta 1`] = `
"import \\"http://example.com/import-meta.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerModule(
    \\"http://example.com/import-meta.mjs\\",
    [],
    []
);"
`;

exports[`generateHmrProxy imports 1`] = `
"import {
    a
} from 'http://example.com/imports.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;

export {
    a0 as a
};

client.registerModule(
    \\"http://example.com/imports.mjs\\",
    [\\"a\\"],
    [\\"http://example.com/a\\",\\"http://example.com/c\\",\\"http://example.com/d\\"],
    (updated) => {
        a0 = updated.a;
    }
);"
`;

exports[`generateHmrProxy imports-only 1`] = `
"import \\"http://example.com/imports-only.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerModule(
    \\"http://example.com/imports-only.mjs\\",
    [],
    [\\"http://example.com/a\\",\\"http://example.com/b\\",\\"http://example.com/c\\"]
);"
`;

exports[`generateHmrProxy multiple-named-exports 1`] = `
"import {
    a,
    b,
    c
} from 'http://example.com/multiple-named-exports.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;
let b0 = b;
let c0 = c;

export {
    a0 as a,
    b0 as b,
    c0 as c
};

client.registerModule(
    \\"http://example.com/multiple-named-exports.mjs\\",
    [\\"a\\",\\"b\\",\\"c\\"],
    [],
    (updated) => {
        a0 = updated.a;
        b0 = updated.b;
        c0 = updated.c;
    }
);"
`;

exports[`generateHmrProxy mutable-binding-let 1`] = `
"export * from \\"http://example.com/mutable-binding-let.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerNonReloadableModule(
    \\"http://example.com/mutable-binding-let.mjs\\"
);"
`;

exports[`generateHmrProxy mutable-binding-var 1`] = `
"export * from \\"http://example.com/mutable-binding-var.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerNonReloadableModule(
    \\"http://example.com/mutable-binding-var.mjs\\"
);"
`;

exports[`generateHmrProxy mutable-destructuring 1`] = `
"export * from \\"http://example.com/mutable-destructuring.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerNonReloadableModule(
    \\"http://example.com/mutable-destructuring.mjs\\"
);"
`;

exports[`generateHmrProxy mutable-named-export 1`] = `
"export * from \\"http://example.com/mutable-named-export.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerNonReloadableModule(
    \\"http://example.com/mutable-named-export.mjs\\"
);"
`;

exports[`generateHmrProxy named-class-export.js 1`] = `
"import {
    C
} from 'http://example.com/named-class-export.js?mtime=0';
import { client } from \\"/@hmr\\";
let C0 = C;

export {
    C0 as C
};

client.registerModule(
    \\"http://example.com/named-class-export.js\\",
    [\\"C\\"],
    [],
    (updated) => {
        C0 = updated.C;
    }
);"
`;

exports[`generateHmrProxy named-conflicts 1`] = `
"import {
    a,
    a0,
    a1,
    a2
} from 'http://example.com/named-conflicts.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a3 = a;
let a00 = a0;
let a10 = a1;
let a20 = a2;

export {
    a3 as a,
    a00 as a0,
    a10 as a1,
    a20 as a2
};

client.registerModule(
    \\"http://example.com/named-conflicts.mjs\\",
    [\\"a\\",\\"a0\\",\\"a1\\",\\"a2\\"],
    [],
    (updated) => {
        a3 = updated.a;
        a00 = updated.a0;
        a10 = updated.a1;
        a20 = updated.a2;
    }
);"
`;

exports[`generateHmrProxy named-export-alias 1`] = `
"import {
    b
} from 'http://example.com/named-export-alias.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let b0 = b;

export {
    b0 as b
};

client.registerModule(
    \\"http://example.com/named-export-alias.mjs\\",
    [\\"b\\"],
    [],
    (updated) => {
        b0 = updated.b;
    }
);"
`;

exports[`generateHmrProxy named-function-declaration 1`] = `
"import {
    f
} from 'http://example.com/named-function-declaration.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let f0 = f;

export {
    f0 as f
};

client.registerModule(
    \\"http://example.com/named-function-declaration.mjs\\",
    [\\"f\\"],
    [],
    (updated) => {
        f0 = updated.f;
    }
);"
`;

exports[`generateHmrProxy named-multiple-var-declarations 1`] = `
"import {
    a,
    b,
    c
} from 'http://example.com/named-multiple-var-declarations.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;
let b0 = b;
let c0 = c;

export {
    a0 as a,
    b0 as b,
    c0 as c
};

client.registerModule(
    \\"http://example.com/named-multiple-var-declarations.mjs\\",
    [\\"a\\",\\"b\\",\\"c\\"],
    [],
    (updated) => {
        a0 = updated.a;
        b0 = updated.b;
        c0 = updated.c;
    }
);"
`;

exports[`generateHmrProxy named-single-export 1`] = `
"import {
    a
} from 'http://example.com/named-single-export.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;

export {
    a0 as a
};

client.registerModule(
    \\"http://example.com/named-single-export.mjs\\",
    [\\"a\\"],
    [],
    (updated) => {
        a0 = updated.a;
    }
);"
`;

exports[`generateHmrProxy named-single-var-declaration 1`] = `
"import {
    a
} from 'http://example.com/named-single-var-declaration.mjs?mtime=0';
import { client } from \\"/@hmr\\";
let a0 = a;

export {
    a0 as a
};

client.registerModule(
    \\"http://example.com/named-single-var-declaration.mjs\\",
    [\\"a\\"],
    [],
    (updated) => {
        a0 = updated.a;
    }
);"
`;

exports[`generateHmrProxy no-exports 1`] = `
"import \\"http://example.com/no-exports.mjs?mtime=0\\";
import { client } from \\"/@hmr\\";

client.registerModule(
    \\"http://example.com/no-exports.mjs\\",
    [],
    []
);"
`;
